src/js/utils/log.js

```diff
@@ -76,32 +83,124 @@ export const logByType = (type, args, stringify = !!IE_VERSION && IE_VERSION < 1
 };
 
 /**
- * Log plain debug messages
+ * Logs plain debug messages. Similar to `console.log`.
  *
- * @param {Mixed[]} args
- *        One or more messages or objects that should be logged.
+ * @class
+ * @param    {Mixed[]} args
+ *           One or more messages or objects that should be logged.
  */
 log = function(...args) {
   logByType('log', args);
 };
 
 /**
- * Keep a history of log messages
+ * Retains a history of all arguments passed to logging functions.
+ *
+ * These records are retained regardless of logging level.
+ *
+ * Whether logging is enabled or not, users would be wise to think carefully
+ * about what they pass to video.js' logging functions. Because values passed
+ * are stored here, they could lead to memory leaks.
  *
  * @type {Array}
  */
 log.history = [];
 
 /**
- * Log error messages
+ * Enumeration of available logging levels, where the keys are the level names
+ * and the values are `|`-separated strings containing logging methods allowed
+ * in that logging level. These strings are used to create a regular expression
+ * matching the function name being called.
+ *
+ * Levels provided by video.js are:
+ *
+ * - `off`: Matches no calls. Any value that can be cast to `false` will have
+ *   this effect. The most restrictive.
+ * - `all` (default): Matches only Video.js-provided functions (`log`,
+ *   `log.warn`, and `log.error`).
+ * - `log`: Matches `log` calls only.
+ * - `warn`: Matches `log.warn` and `log.error` calls.
+ * - `error`: Matches only `log.error` calls.
+ *
+ * @type {Object}
+ */
+log.levels = {};
+
+if (Object.defineProperties) {
+  Object.defineProperties(log.levels, {
+    all: {
+      enumerable: true,
+      value: 'log|warn|error'
```

Comment:
typical log level implementations that I have seen use number rather rather than regex matches to see if they should log. I think that this works fine here and it does allow us to only log from one level if we want to, but I figured that I would throw in another idea.

Here are the log levels I see:
* 0 - Off
* 1 - Error
* 2 - Warn
* 3 - log

Typically you set your log level to the number or string representation of a level, so 3 or `log` by default. Then everything below that level is logged. So if the log level is 3 we log everything, if the log level is 2 we log only warnings and errors, if the log level is 1 we only log errors, and finally if the log level is 0 we log nothing at all.